peak/float.py

```diff
@@ -1,38 +1,424 @@
 from functools import lru_cache
 from types import SimpleNamespace
-from . import Peak, family_closure
-from hwtypes import BitVector
-from hwtypes import TypeFamily
-from .family import BlackBox
+
+from peak import family_closure, Peak
+from peak.family import SMTFamily, MagmaFamily
+from hwtypes import SMTFPVector, FPVector, RoundingMode as RM, TypeFamily
+from peak.black_box import BlackBox
+from hwtypes import BitVector, Bit
+from hwtypes.adt import Enum
+import magma
+
+class RoundingMode(Enum):
+    RNE = 0
+    RTZ = 1
+    RDN = 2
+    RUP = 3
+    RMM = 4
+
+def RoudningMode_utils(family):
+    RM = RoundingMode
+    RM_c = family.get_constructor(RoundingMode)
+    BV3 = lambda e: family.BitVector[3](e.value)
+    def from_rm(rm: RoundingMode) -> BitVector[3]:
+        return (rm==RM.RNE).ite(
+            BV3(RM.RNE),
+            (rm==RM.RTZ).ite(
+                BV3(RM.RTZ),
+                (rm == RM.RDN).ite(
+                    BV3(RM.RDN),
+                    (rm == RM.RUP).ite(
+                        BV3(RM.RUP),
+                        BV3(RM.RMM),
+                    ),
+                ),
+            ),
+        )
+
+    def to_rm(rm: BitVector[3]) -> RoundingMode:
+        return (rm==BV3(RM.RNE)).ite(
+            RM_c(RM.RNE),
+            (rm==BV3(RM.RTZ)).ite(
+                RM_c(RM.RTZ),
+                (rm == BV3(RM.RDN)).ite(
+                    RM_c(RM.RDN),
+                    (rm == BV3(RM.RUP)).ite(
+                        RM_c(RM.RUP),
+                        RM_c(RM.RMM),
+                    ),
+                ),
+            ),
+        )
+
+    return SimpleNamespace(**locals())
 
 @lru_cache(None)
-def Float(frac, exp):
-    width = frac + exp + 1
+def float_lib_gen(exp_bits: int, frac_bits: int):
```

Comment:
Why do some of the operations (`add`, `sub`, ...) have implementations while others do not (`neg`, `abs`, ...)?

Reply:
I'd prefer they all have impls.

Reply:
add/mul are the ones that really needed implementations due to Lassen. The other ones I will implement lazily when they become needed.